File: Samples/RadialMenu/UE4/Source/RadialMenu/Arc.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "Arc.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/UIElementData.h>
#include <NsGui/FrameworkPropertyMetadata.h>
#include <NsGui/Brush.h>
#include <NsGui/Path.h>
#include <NsGui/StreamGeometry.h>
#include <NsGui/StreamGeometryContext.h>
#include <NsGui/Binding.h>


using namespace RadialMenu;


////////////////////////////////////////////////////////////////////////////////////////////////////
Arc::Arc()
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Arc::~Arc()
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetCenterX() const
{
    return GetValue<float>(CenterXProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetCenterX(float value)
{
    SetValue<float>(CenterXProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetCenterY() const
{
    return GetValue<float>(CenterYProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetCenterY(float value)
{
    SetValue<float>(CenterYProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetInnerRadius() const
{
    return GetValue<float>(InnerRadiusProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetInnerRadius(float value)
{
    SetValue<float>(InnerRadiusProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetOuterRadius() const
{
    return GetValue<float>(OuterRadiusProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetOuterRadius(float value)
{
    SetValue<float>(OuterRadiusProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetOffset() const
{
    return GetValue<float>(OffsetProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetOffset(float value)
{
    SetValue<float>(OffsetProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetAngleStart() const
{
    return GetValue<float>(AngleStartProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetAngleStart(float value)
{
    SetValue<float>(AngleStartProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetAngleDelta() const
{
    return GetValue<float>(AngleDeltaProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetAngleDelta(float value)
{
    SetValue<float>(AngleDeltaProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetPadding() const
{
    return GetValue<float>(PaddingProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetPadding(float value)
{
    SetValue<float>(PaddingProperty, value);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::Brush* Arc::GetFill() const
{
    return GetValue<Noesis::Ptr<Noesis::Brush>>(FillProperty);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetFill(Noesis::Brush* value)
{
    SetValue<Noesis::Ptr<Noesis::Brush>>(FillProperty, value);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::Brush* Arc::GetStroke() const
{
    return GetValue<Noesis::Ptr<Noesis::Brush>>(StrokeProperty);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetStroke(Noesis::Brush* value)
{
    SetValue<Noesis::Ptr<Noesis::Brush>>(StrokeProperty, value);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
float Arc::GetStrokeThickness() const
{
    return GetValue<float>(StrokeThicknessProperty);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::SetStrokeThickness(float value)
{
    SetValue<float>(StrokeThicknessProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::Size Arc::MeasureOverride(const Noesis::Size& availableSize)
{
    BuildArcGeometry();
    return ParentClass::MeasureOverride(availableSize);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
static Noesis::Point RadialPoint(Noesis::Point center, float angle, float radius)
{
    float radiansAngle = (angle - 90.0f) * Noesis::DegToRad;
    float x = radius * cosf(radiansAngle);
    float y = radius * sinf(radiansAngle);
    return Noesis::Point(x + center.x, y + center.y);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::BuildArcGeometry()
{
    EnsurePath();

    Noesis::Path* path = (Noesis::Path*)GetChild();
    Noesis::StreamGeometry* geometry = (Noesis::StreamGeometry*)path->GetData();
    Noesis::StreamGeometryContext geo = geometry->Open();

    float angleDelta = GetAngleDelta();
    float outerRadius = GetOuterRadius();
    float innerRadius = GetInnerRadius();
    if (angleDelta > 0 && fabsf(outerRadius - innerRadius) > FLT_EPSILON)
    {
        float angleStart = GetAngleStart();
        float outerStartAngle = angleStart;
        float outerAngleDelta = angleDelta;
        float innerStartAngle = angleStart;
        float innerAngleDelta = angleDelta;
        float offset = GetOffset();
        float padding = GetPadding();
        Noesis::Point center(GetCenterX(), GetCenterY());
        Noesis::Size outerSize(outerRadius, outerRadius);
        Noesis::Size innerSize(innerRadius, innerRadius);

        // If we have to draw a full-circle, we do it with two semi-circles, because 'ArcTo()' can
        // not draw a full-circle
        if (angleDelta >= 360.0f && padding <= 0.0f)
        {
            Noesis::Point outerTop = RadialPoint(center, outerStartAngle, outerRadius + offset);
            Noesis::Point outerBottom = RadialPoint(center, outerStartAngle + 180.0f, outerRadius + offset);
            Noesis::Point innerTop = RadialPoint(center, innerStartAngle, innerRadius + offset);
            Noesis::Point innerBottom = RadialPoint(center, innerStartAngle + 180.0f, innerRadius + offset);

            geo.BeginFigure(innerTop, true);
            geo.LineTo(outerTop);
            geo.ArcTo(outerBottom, outerSize, 0.0f, false, Noesis::SweepDirection_Clockwise);
            geo.ArcTo(outerTop, outerSize, 0.0f, false, Noesis::SweepDirection_Clockwise);
            geo.LineTo(innerTop);
            geo.ArcTo(innerBottom, innerSize, 0.0f, false, Noesis::SweepDirection_Counterclockwise);
            geo.ArcTo(innerTop, innerSize, 0.0f, false, Noesis::SweepDirection_Counterclockwise);
        }
        else
        {
            if (padding > FLT_EPSILON)
            {
                // Offsets the angle by the padding
                float outerAngleVariation = (180.0f * (padding / outerRadius)) / Noesis::Pi;
                float innerAngleVariation = (180.0f * (padding / innerRadius)) / Noesis::Pi;

                outerStartAngle += outerAngleVariation;
                outerAngleDelta -= outerAngleVariation * 2.0f;
                innerStartAngle += innerAngleVariation;
                innerAngleDelta -= innerAngleVariation * 2.0f;
            }

            Noesis::Point outerStart = RadialPoint(center, outerStartAngle, outerRadius + offset);
            Noesis::Point outerEnd = RadialPoint(center, outerStartAngle + outerAngleDelta, outerRadius + offset);
            Noesis::Point innerStart = RadialPoint(center, innerStartAngle, innerRadius + offset);
            Noesis::Point innerEnd = RadialPoint(center, innerStartAngle + innerAngleDelta, innerRadius + offset);

            bool largeArcOuter = outerAngleDelta > 180.0f;
            bool largeArcInner = innerAngleDelta > 180.0f;

            geo.BeginFigure(innerStart, true);
            geo.LineTo(outerStart);
            geo.ArcTo(outerEnd, outerSize, 0.0f, largeArcOuter, Noesis::SweepDirection_Clockwise);
            geo.LineTo(innerEnd);
            geo.ArcTo(innerStart, innerSize, 0.0f, largeArcInner, Noesis::SweepDirection_Counterclockwise);
        }
    }

    geo.Close();

    path->InvalidateMeasure();
    path->InvalidateVisual();
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void Arc::EnsurePath()
{
    if (GetChild() == nullptr)
    {
        Noesis::Ptr<Noesis::StreamGeometry> geometry = *new Noesis::StreamGeometry();
        geometry->SetFillRule(Noesis::FillRule_EvenOdd);

        Noesis::Ptr<Noesis::Path> path = *new Noesis::Path();
        path->SetData(geometry);

        path->SetBinding(Noesis::Shape::FillProperty, Noesis::MakePtr<Noesis::Binding>(FillProperty, this));
        path->SetBinding(Noesis::Shape::StrokeProperty, Noesis::MakePtr<Noesis::Binding>(StrokeProperty, this));
        path->SetBinding(Noesis::Shape::StrokeThicknessProperty, Noesis::MakePtr<Noesis::Binding>(StrokeThicknessProperty,
            this));

        SetChild(path);
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(Arc, "RadialMenu.Arc")
{
    Noesis::UIElementData* data = NsMeta<Noesis::UIElementData>(Noesis::TypeOf<SelfClass>());

    auto onShapePropertyChanged = [](Noesis::DependencyObject* d, const Noesis::DependencyPropertyChangedEventArgs&)
    {
        ((Arc*)d)->InvalidateMeasure();
    };

    data->RegisterProperty<float>(CenterXProperty, "CenterX",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(CenterYProperty, "CenterY",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(InnerRadiusProperty, "InnerRadius",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(OuterRadiusProperty, "OuterRadius",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(OffsetProperty, "Offset",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(AngleStartProperty, "AngleStart",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(AngleDeltaProperty, "AngleDelta",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<float>(PaddingProperty, "Padding",
        Noesis::FrameworkPropertyMetadata::Create(0.0f, Noesis::PropertyChangedCallback(onShapePropertyChanged)));
    data->RegisterProperty<Noesis::Ptr<Noesis::Brush>>(FillProperty, "Fill",
        Noesis::FrameworkPropertyMetadata::Create(Noesis::Ptr<Noesis::Brush>()));
    data->RegisterProperty<Noesis::Ptr<Noesis::Brush>>(StrokeProperty, "Stroke",
        Noesis::FrameworkPropertyMetadata::Create(Noesis::Ptr<Noesis::Brush>()));
    data->RegisterProperty<float>(StrokeThicknessProperty, "StrokeThickness",
        Noesis::FrameworkPropertyMetadata::Create(1.0f));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Noesis::DependencyProperty* Arc::CenterXProperty;
const Noesis::DependencyProperty* Arc::CenterYProperty;
const Noesis::DependencyProperty* Arc::InnerRadiusProperty;
const Noesis::DependencyProperty* Arc::OuterRadiusProperty;
const Noesis::DependencyProperty* Arc::OffsetProperty;
const Noesis::DependencyProperty* Arc::AngleStartProperty;
const Noesis::DependencyProperty* Arc::AngleDeltaProperty;
const Noesis::DependencyProperty* Arc::PaddingProperty;
const Noesis::DependencyProperty* Arc::FillProperty;
const Noesis::DependencyProperty* Arc::StrokeProperty;
const Noesis::DependencyProperty* Arc::StrokeThicknessProperty;

File: Samples/RadialMenu/UE4/Assets/RadialMenu/RadialMenu/Arc.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Shapes;

namespace RadialMenu
{
    /// <summary>
    /// Panel to display items in a radial/arc.
    /// </summary>
    public class Arc : Decorator
    {
        #region Dependency Properties
        /// <summary>
        /// The center X coordinate of the circle from which this arc piece is cut
        /// </summary>
        public double CenterX
        {
            get { return (double)GetValue(CenterXProperty); }
            set { SetValue(CenterXProperty, value); }
        }

        public static readonly DependencyProperty CenterXProperty = DependencyProperty.Register(
            "CenterX", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The center Y coordinate of the circle from which this arc piece is cut
        /// </summary>
        public double CenterY
        {
            get { return (double)GetValue(CenterYProperty); }
            set { SetValue(CenterYProperty, value); }
        }

        public static readonly DependencyProperty CenterYProperty = DependencyProperty.Register(
            "CenterY", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The inner radius of this arc piece
        /// </summary>
        public double InnerRadius
        {
            get { return (double)GetValue(InnerRadiusProperty); }
            set { SetValue(InnerRadiusProperty, value); }
        }

        public static readonly DependencyProperty InnerRadiusProperty = DependencyProperty.Register(
            "InnerRadius", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The outer radius of this arc piece
        /// </summary>
        public double OuterRadius
        {
            get { return (double)GetValue(OuterRadiusProperty); }
            set { SetValue(OuterRadiusProperty, value); }
        }

        public static readonly DependencyProperty OuterRadiusProperty = DependencyProperty.Register(
            "OuterRadius", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The distance to translate this arc piece out from the center
        /// </summary>
        public double Offset
        {
            get { return (double)GetValue(OffsetProperty); }
            set { SetValue(OffsetProperty, value); }
        }

        public static readonly DependencyProperty OffsetProperty = DependencyProperty.Register(
            "Offset", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The start angle from the Y axis vector of this arc piece in degrees
        /// </summary>
        public double AngleStart
        {
            get { return (double)GetValue(AngleStartProperty); }
            set { SetValue(AngleStartProperty, value); }
        }

        public static readonly DependencyProperty AngleStartProperty = DependencyProperty.Register(
            "AngleStart", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The angle delta from the Y axis vector of this arc piece in degrees
        /// </summary>
        public double AngleDelta
        {
            get { return (double)GetValue(AngleDeltaProperty); }
            set { SetValue(AngleDeltaProperty, value); }
        }

        public static readonly DependencyProperty AngleDeltaProperty = DependencyProperty.Register(
            "AngleDelta", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        /// <summary>
        /// The padding around this arc piece
        /// </summary>
        public double Padding
        {
            get { return (double)GetValue(PaddingProperty); }
            set { SetValue(PaddingProperty, value); }
        }

        public static readonly DependencyProperty PaddingProperty = DependencyProperty.Register(
            "Padding", typeof(double), typeof(Arc), new FrameworkPropertyMetadata(0.0, OnShapePropertyChanged));

        private static void OnShapePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            ((Arc)d).InvalidateMeasure();
        }

        /// <summary>
        /// Brush that fills the arc
        /// </summary>
        public Brush Fill
        {
            get { return (Brush)GetValue(FillProperty); }
            set { SetValue(FillProperty, value); }
        }

        public static readonly DependencyProperty FillProperty = DependencyProperty.Register(
            "Fill", typeof(Brush), typeof(Arc), new PropertyMetadata(null));

        /// <summary>
        /// Brush used to stroke the arc
        /// </summary>
        public Brush Stroke
        {
            get { return (Brush)GetValue(StrokeProperty); }
            set { SetValue(StrokeProperty, value); }
        }

        public static readonly DependencyProperty StrokeProperty = DependencyProperty.Register(
            "Stroke", typeof(Brush), typeof(Arc), new PropertyMetadata(null));

        /// <summary>
        /// The thickness of the stroke applied to the arc
        /// </summary>
        public double StrokeThickness
        {
            get { return (double)GetValue(StrokeThicknessProperty); }
            set { SetValue(StrokeThicknessProperty, value); }
        }

        public static readonly DependencyProperty StrokeThicknessProperty = DependencyProperty.Register(
            "StrokeThickness", typeof(double), typeof(Arc), new PropertyMetadata(1.0));
        #endregion

        static Arc()
        {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(Arc), new FrameworkPropertyMetadata(typeof(Arc)));
        }

        protected override Size MeasureOverride(Size constraint)
        {
            BuildArcGeometry();

            return base.MeasureOverride(constraint);
        }

        private void EnsurePath()
        {
            if (Child == null)
            {
                Path path = new Path() { Data = new StreamGeometry { FillRule = FillRule.EvenOdd } };
                path.SetBinding(Shape.FillProperty, new Binding("Fill") { Source = this });
                path.SetBinding(Shape.StrokeProperty, new Binding("Stroke") { Source = this });
                path.SetBinding(Shape.StrokeThicknessProperty, new Binding("StrokeThickness") { Source = this });

                Child = path;
            }
        }

        private void BuildArcGeometry()
        {
            EnsurePath();

            StreamGeometry geometry = ((Path)Child).Data as StreamGeometry;

            using (StreamGeometryContext context = geometry.Open())
            {
                if (AngleDelta > 0 && Math.Abs(OuterRadius - InnerRadius) > double.Epsilon)
                {
                    double outerStartAngle = AngleStart;
                    double outerAngleDelta = AngleDelta;
                    double innerStartAngle = AngleStart;
                    double innerAngleDelta = AngleDelta;
                    Point center = new Point(CenterX, CenterY);
                    Size outerArcSize = new Size(OuterRadius, OuterRadius);
                    Size innerArcSize = new Size(InnerRadius, InnerRadius);

                    // If we have to draw a full-circle, we do it with two semi-circles, because 'ArcTo()' can
                    // not draw a full-circle
                    if (AngleDelta >= 360 && Padding <= 0)
                    {
                        Point outerTop = RadialPoint(center, outerStartAngle, OuterRadius + Offset);
                        Point outerBottom = RadialPoint(center, outerStartAngle + 180, OuterRadius + Offset);
                        Point innerTop = RadialPoint(center, innerStartAngle, InnerRadius + Offset);
                        Point innerBottom = RadialPoint(center, innerStartAngle + 180, InnerRadius + Offset);

                        context.BeginFigure(innerTop, true, true);
                        context.LineTo(outerTop, true, true);
                        context.ArcTo(outerBottom, outerArcSize, 0, false, SweepDirection.Clockwise, true, true);
                        context.ArcTo(outerTop, outerArcSize, 0, false, SweepDirection.Clockwise, true, true);
                        context.LineTo(innerTop, true, true);
                        context.ArcTo(innerBottom, innerArcSize, 0, false, SweepDirection.Counterclockwise, true, true);
                        context.ArcTo(innerTop, innerArcSize, 0, false, SweepDirection.Counterclockwise, true, true);
                    }
                    else
                    {
                        if (Padding > 0)
                        {
                            // Offsets the angle by the padding
                            double outerAngleVariation = (180 * (Padding / OuterRadius)) / Math.PI;
                            double innerAngleVariation = (180 * (Padding / InnerRadius)) / Math.PI;

                            outerStartAngle += outerAngleVariation;
                            outerAngleDelta -= outerAngleVariation * 2;
                            innerStartAngle += innerAngleVariation;
                            innerAngleDelta -= innerAngleVariation * 2;
                        }

                        Point outerStart = RadialPoint(center, outerStartAngle, OuterRadius + Offset);
                        Point outerEnd = RadialPoint(center, outerStartAngle + outerAngleDelta, OuterRadius + Offset);
                        Point innerStart = RadialPoint(center, innerStartAngle, InnerRadius + Offset);
                        Point innerEnd = RadialPoint(center, innerStartAngle + innerAngleDelta, InnerRadius + Offset);

                        bool largeArcOuter = outerAngleDelta > 180.0;
                        bool largeArcInner = innerAngleDelta > 180.0;

                        context.BeginFigure(innerStart, true, true);
                        context.LineTo(outerStart, true, true);
                        context.ArcTo(outerEnd, outerArcSize, 0, largeArcOuter, SweepDirection.Clockwise, true, true);
                        context.LineTo(innerEnd, true, true);
                        context.ArcTo(innerStart, innerArcSize, 0, largeArcInner, SweepDirection.Counterclockwise, true, true);
                    }
                }
            }
        }

        private static Point RadialPoint(Point center, double angle, double radius)
        {
            // Converts to radians
            double radiansAngle = (angle - 90) * Math.PI / 180.0;
            double x = radius * Math.Cos(radiansAngle);
            double y = radius * Math.Sin(radiansAngle);
            return new Point(x + center.X, y + center.Y);
        }
    }
}
